File: Interviews/GFG/random-problems/Medium/Filling-Bucket/Filling-Bucket.py
class Solution:
    def fillingBucket(self, N):
        cur, pre, prepre, mod = 0, 0, 0, int(1e8)
        for i in range(N + 1):
            if i == 0: cur = 1
            else:
                cur = (cur + pre) % mod
                if i > 1:
                    cur = (cur + prepre) % mod
            prepre = pre
            pre = cur
            cur = 0
        
        return pre

File: Interviews/GFG/random-problems/Medium/Filling-Bucket/Filling-Bucket.cpp
class Solution {
public:
    int fillingBucket(int N, int mod = 1e8) {
        vector<int> cache(N + 1, 0);
        for (int i = 0; i <= N; i++) {
            if (i == 0) {
                cache[i] = 1;
            } else {
                cache[i] = (cache[i] + cache[i - 1]) % mod;
                if (i > 1) {
                    cache[i] = (cache[i] + cache[i - 2]) % mod;
                }
            }
        }

        return cache[N];
    }
};